hazelcast/src/hazelcast/client/connection/InSelector.cpp

```diff
@@ -66,7 +66,7 @@ namespace hazelcast {
                             sleepingSocket->receive(&wakeUpSignal, sizeof(int));
                         } else {
                             boost::shared_ptr<Connection> conn = connectionManager.getConnectionIfAvailable(fd);
-                            if (conn.get() != NULL) {
+                            if (conn.get() != NULL && conn->live) {
```

Comment:
Connection live is closed and then it is removed from this list. 
As long as we don't change this order, it is possible that connection is closed after your `live` check here. 
It think fix should be fixing the order of 
`socket->close();`
with followings . 
```
socketConnections.remove(socketId);
connections.remove(address);
```